java/org/apache/catalina/realm/RealmBase.java

```diff
@@ -928,6 +928,15 @@ public boolean hasRole(Wrapper wrapper, Principal principal, String role) {
             }
         }
 
+        // Check for a role alias/mapping defined on context level
+        if (getContainer() instanceof Context) {
+            Context context = (Context) getContainer();
+            String realRole = context.findRoleMapping(role);
```

Comment:
The above code is only used when the Realm is defined at the Context level. That isn't what is required here. The role mappings need to be checked for the current Context irrespective of where the Realm is defined. Something like:
`Context context = (Context) wrapper.getParent();`

A similar change would also need to be made to `UserDatabaseRealm` and potentially other sub-classes. Consider moving this to `Wrapper.findSecurityReference`. i.e. Look up Wrapper references first and if none found try the Context.

Reply:
So, if I understand your request properly, this should not be part of `RealmBase` at all, but has to be implemented in `StandardWrapper`, namely `findSecurityReference()` whre it queries `getParent()#findRoleMapping() if it is not found in `references`?.

So `findSecurityReferences()` have to be changed as well? But not `removeSecurityReference()`?

Reply:
`findSecurityReferences()` should be unchanged.

Parent of a Wrapper should only ever be a Context but an `instance of` check is probably prudent. It would also protect against nulls (may be possible during shutdown - I didn't check).

Reply:
Both

> Consider moving this to Wrapper.findSecurityReference

and

 > findSecurityReferences() should be unchanged.

seem to be a contradiction. Do you want me to change the `Wrapper` interface and all of its implementors? That would still mean that I need to change `RealmBase` anyway?!

Reply:
Just take your time to describe how you image that, I do not fully understand it.

Reply:
`findSecurityReference()` != `findSecurityReferences()`

`findSecurityReference()` is the method that tries to map a single role and should be changed.
`findSecurityReferences()` is closer to a getter for the current settings for the Wrapper and should be unchanged.

So:
```
@Override
public String findSecurityReference(String name) {
    String reference;
        
    // First check the Wrapper 
    referencesLock.readLock().lock();
    try {
        reference = references.get(name);
    } finally {
        referencesLock.readLock().unlock();
    }

    // If not specified on the Wrapper, check the Context
    if (reference == null && getParent() instanceof Context) {
        reference = ((Context) getParent()).findRoleMapping(name);
    }
        
    return reference;
}
```

Reply:
Wouldn't it be confusing when `findSecurityReference()` can map a role which `findSecurityReferences()` does not contain?

Why the null check? Wouldn't it be possible for a servlet security role ref to be also mapped to a technical role? My current does this.

Reply:
@markt-asf I have added another commit with the requested change. It that meets your expectations, I will modify the tests accordingly.

Reply:
You can argue for and against the null check. If you prefer to skip it, I'm fine with that.

Reply:
I am not really skipping it, I have pushed into the check loop. You previous check ignored a valid case.